### DISCUSSION TITLE
Auto-completion for import-statements (file lookup in workspace)

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1104] [Q&A] [answered]
Is there an example/description on how to implement a auto completion for an import statement in a DSL (like the DSL in #979).

--> I would like to get a suggestion while typing `import "./c<CTRL-SPACE>"

I started with a custom `CompletionProvider`, but I am not sure if this is the way to go (in this provider I could lookup the workspace and compute all relative file locations).

**Comment by msujew** [comment]
We have a long-standing issue that tries to address a similar issue in the grammar DSL, see https://github.com/langium/langium/issues/350.

Ideally, you should just override the `completionFor` method in the `CompletionProvider`. You need to dispatch on the correct feature and apply your custom completion logic that performs the mentioned lookup in the workspace.

However, I've just noticed an issue that crops up related to completion in strings, which doesn't seem to work at all right now. It works when you start the completion before the `"` though.

**Comment by goto40** [comment]
Yes, I confirm (completion starts when triggered before "). It seems I get a double quote at the beginning...

Thank you for the prompt response. I will work with this until we have a better solution.

**Reply by msujew** [reply, marked as answer]
See https://github.com/langium/langium/pull/1106 for an example of how to accomplish import completion. The PR fixes a few minor issues in the base completion provider related to completions inside of strings (or other, similar tokens).

**Reply by goto40** [reply]
Thank you. I took some ideas - but the complete thing did not work for me (I had problems, when completing a string with a dot `".<CTRL-SPACE>`  - then my completion handler was not triggered correctly...

**Reply by msujew** [reply]
Yeah, that much is expected. If you take a look at the PR, you'll see that it changes some stuff in the completion provider, which are necessary for this to work. We'll release a new version with these fixes soon.

**Reply by goto40** [reply]
WIth https://github.com/eclipse-langium/langium/pull/1106/files#diff-01998089cdf0366c9bb26797a1875e637705794708a674e18738f408b155e308R39 auto completion for this usecase works as expected! Thank you.
